specification/dartLangSpec.tex

```diff
@@ -11496,6 +11506,8 @@ \subsection{Assignable Expressions}
 
 <assignableSelector> ::= <unconditionalAssignableSelector>
   \alt `?.' <identifier>
+  \alt `!'
```

Comment:
> I don't think `x!` should be _assignable_. Put it into `<selector>` instead.

If you do this, then you can't write `x!.y = 3`, which I think we want to be able to write.  Unless I'm missing something?  But I think the `x!.y` can't be derived as a selector, you have to go through assignableSelector.  So I was planning on adding this here, and then disallowing things like `x! = 3` semantically.  

> You can write garbage like `x..foo() = 42` with the existing grammar too,

How does this derive?  I don't see how `x..foo()` gets derived as an `<assignableExpression>`.

Reply:
If you change `selector` to:
```ebnf
<selector> ::= `!'
 \alt <argumentPart>
 \alt <assignableSelector>
```
*and* either `assignableSelectorPart` to:
```
<assignableSelectorPart> ::= (<argumentPart> | `!')* <assignableSelector>
```
(from `<assignableSelectorPart> ::= <argumentPart>* <assignableSelector>`)
or change that and `assignableExpression` to:
```ebnf
<assignableeExpression> ::= <primary> <assignableSelectorPart> 
  \alt \SUPER{} <unconditionalAssignableSelector>
  \alt <constructorInvocation> <assignableSelectorPart>
  \alt <identifier>

<assignableSelectorPart> ::= <selector>* <assignableSelector>
```
then you can derive `x!.y = 3`.

You are correct that just changing `<selector>` si not enough because `assignableSelectorPart` doesn't use `selector`.

---

About `..foo() = 42`, the grammar for `cascadeSection` is
```ebnf
<cascadeSection> ::= `..' (<cascadeSelector> <argumentPart>*)
  \gnewline{} (<assignableSelector> <argumentPart>*)*
  \gnewline{} (<assignmentOperator> <expressionWithoutCascade>)?
```
and with:
 *  `<cascadeSelector>` &mapsto; `<identifier>` &mapsto; `foo`
 * `<argumentPart>` &mapsto; `<arguments>` &mapsto; `(` `)`
 * `<assignmentOperator>` &mapsto; `=`
 * `<expressionWithoutCascade>` &mapsto;<sup>*</sup> `42`

you can derive `x .. foo () = 42` by having one `<argumentPart>`, zero `(<assignableSelector> <argumentPart>*)` and one `(<assignmentOperator> <expressionWthoutCascade>)` as allowed by the multiplicities.
There is no `<assignableExpression>` involved.

The problem is that `<argumentPart>` is a selector and not an assignable selector, but the cascade grammar allows inserting it before the assignment. The issue is the cascade grammar, and as you say the `<assignableExpression>` doesn't have that issue.
I think my suggestion above is better (but would obviously like someone else to check that it still allows all the right things).

The code is rejected by our tools anyway, even though it satisfies the grammar. It is then treated as equivalent to `x.foo() = 42`, which is itself invalid. Example:
```dart
main() {
  42..toString() = 4;
} 
```
gives the CFE error:
```
casc.dart:2:18: Error: Can't assign to this.
  42..toString() = 4;
```
and the analyzer error:
```
Analyzing casc.dart...
  error • Missing selector such as '.<identifier>' or '[0]' at casc.dart:2:5 • missing_assignable_selector
  error • A value of type 'int' can't be assigned to a variable of type 'String' at casc.dart:2:20 • invalid_assignment
2 errors found.
```
(The second error is spurious—there is no variable of type string in the program—which suggests that the analyzer isn't entirely comfortable with this misuse of syntax).

Reply:
@lrhn If I do as you suggest, then I don't think you can have `!` in a cascade section anymore, right?

```
a..foo()!.bar=3;
```

I think this cascade parses with my grammar as:

```
cascadeSelector -> identifier -> foo
argumentPart* -> argumentPart -> ()
assignableSelector -> !
assignableSelector -> .bar
assignmentOperator -> =
expressionWithoutCascade -> conditionalExpression -> ifNullExpression -> logicalOrExpression -> logicalAndExpression -> equalityExpression -> relationExpression -> bitwiseOrExpression -> bitwiseXOrExpression -> bitwiseAndExpression -> shiftExpression -> additiveExpression -> multiplicativeExpression -> unaryExpression -> StayWithMeAlmostThere -> primary -> literal -> numericLiteral
```

How does it parse with yours?

Reply:
Let's pick the first grammar I proposed:
```ebnf
<cascadeSequence> ::= (`?..' \alt `..') <cascadeSection> (`..' <cascadeSection>)*

<cascadeSection> ::= <cascadeSelector> (<cascadeAssignment> | <selector>* (<assignableSelector> <cascadeAssignment>)?)

<selector> ::= `!' 
  \alt <argumentPart>
  \alt <assignableSelector>

<assignableSelector> ::= (`?.' \alt `.') <identifier>
  \alt `?.'? `[' expression `]'

<cascadeAssignment> ::= <assignmentOperator> <expressionWithoutCascade>
```
Then the cascade section of
```dart
a...foo()!.bar = 3
```
parses as
* cascadeSection &mapsto; cascadeSelector selector* assignableSelector cascadeAssignment
* cascadeSelector &mapsto; identifier (`foo`)
* selector* &mapsto; selector selector
* selector &mapsto; argumentPart &mapsto;* `(` `)`
* selector &mapsto; `!`
* assignableSelector &mapsto; `.` identifier (`bar`)
* cascadeAssignment &mapsto; `=` expressionWithoutCascade
* expressionWithoutCascade &mapsto;* integerLiteral (3)

We only want the actually assignable selectors to be assignable, and `!` does not introduce a LHS, so it should not be an assignable selector. Only `.id`, `[e]` and `?.[e]` are assignable because it's the only suffixes that can denote a setter (counting `operator[]=` as a settter).

Reply:
Okay, I've taken your suggestion, PTAL.